### ISSUE TITLE
Enable electron process restarting on file change

### ISSUE THREAD

**Issue by catc** [adopted-ember-addons/ember-electron#283]
Is there any way to restart the electron server when changing something in electron process (ie: `ember-electron/main.js`), as opposed to having to stop the entire ember-cli process, restart it, and wait for ember to build again before being able to test out electron code?

Or possibly have the option to start the electron main process as a separate process from ember-cli?

**Comment by anulman** [comment]
@catc I think we would be happy to entertain proposals or further discussion on this, but I don't think we have enough to act on now. Can you please explain what issues you are experiencing, or nuisances you are trying to resolve? Do you have a proposed architecture in mind?

Overall, I think this will be difficult to implement without surprising e-electron users. For example, we will still need to reboot the entire app—naively on _any_ change, or otherwise by statically determining whether the changed code paths are / would be hit during app initialization. Either way, I (personally!) think it is better for the developer to opt into app (process) reboots, vs have them happen automatically; there is no guarantee proper process cleanup code is in place yet.

That said, I think we're all willing to learn more about your use case, or explore any architecture proposals you might put forward. Will tag this with `more-info-needed` for now, but we will likely close this issue by Jan 2018 unless further discussion is warranted.

**Comment by wycats** [comment]
@anulman or whoever's still watching: this seems like a pretty straight forward request. When making changes to `main.js`, you currently need to restart Ember CLI to see the change. This takes a while, and it would be nicer if the server could restart in the background.

**Comment by anulman** [comment]
@wycats I've not been involved with an Electron project for 18+ months, so unfortunately would be the wrong person to move this forward. Interested parties should look at closing and restarting the Electron handle returned by [this invocation of `electron-forge`'s start command](https://github.com/adopted-ember-addons/ember-electron/blob/master/lib/commands/electron.js#L86) on relevant file changes.

I do not recall if ember-electron is listening for changes in `main.js` or its imports. If so they would be registered via [these invocations of the `Builder` and `Watcher` constructors](https://github.com/adopted-ember-addons/ember-electron/blob/master/lib/commands/electron.js#L54-L67); if not they would need to be watched separately.

**Comment by anulman** [comment]
Also I should note that while my earlier ignorance of Ember CLI build primitives overcomplicated the matter, a point from my prior comment still stands: having the app restart in the background may have serious ramifications when said app is not sandboxed in a browser tab.

For example, one app I previously worked on mounted a FUSE filesystem at boot; were the app to restart in the background what sort of cleanup should be done? Assuming the Ember BrowserWindow was open, how do we restore state? Should we?

**Comment by bendemboski** [comment]
I suspect it wouldn't be too difficult to modify that `_startElectron()` method (that Aidan linked above) to monitor some files and restart, but the question of which files to monitor and how is a slightly tricky one. Certainly the `ember-electron` directory, but that would ignore any changes to `node_modules`, and there's no good way of knowing which dependencies affect the Ember build and which affect the main process build, so I think we'd have to just ignore `node_modules` or else risk restarting the whole app when an ember rebuild/window reload is all that's needed.

Responding to changes in `node_modules` is probably of secondary priority to responding to changes the the application's main process code, so maybe it's okay to just ignore them.

I doubt I would want to use this feature myself as my application is a document editor, and prompts to save on quit, allowing some user interaction before the application actually quits (and allowing the user to abort the quit), and also keeps data in temp files when the user hasn't saved it out to a file. So, like Aidan's app, restarting my app is a pretty significant operation that I want under my direct control, and a naive implementation of the app restart behavior would often and easily leave me with multiple copies of the app running (some stuck on user input in their quit handlers).

So I'd advocate for this feature being opt-in rather than opt-out, at least initially. If somebody wants to take a stab at it to see how it actually works out, I'd be happy to support the effort with my knowledge/familiarity with the code, but I'm not planning to make time to do this myself in the foreseeable future.

**Comment by jacobq** [comment]
> So I'd advocate for this feature being opt-in rather than opt-out, at least initially. If somebody wants to take a stab at it to see how it actually works out, I'd be happy to support the effort with my knowledge/familiarity with the code, but I'm not planning to make time to do this myself in the foreseeable future.

I am of the same opinion. My biggest electron app uses the main thread to control physical devices. If it were to get killed ungracefully then it could potentially leave sensitive things in a bad state (e.g. pumps, motors, high-voltage field generators, etc. -- admittedly, safer/cheaper "fake" equipment is usually substituted in development, but even so, in this scenario underlying connections may have to timeout before the main thread can successfully start again). 

For this feature to be useful to me it would need to incorporate some kind of signalling to the main program to say, "We're going to shutdown. Let me know when you're OK with that. I'll give you up to 10 seconds if I don't hear from you."

**Comment by bendemboski** [comment]
For anyone still interested in this, the story is somewhat improved in v3 (beta out now). If you run `ember electron:build` it will build the Ember app into your `electron-app/ember-dist` folder. Your `electron-app` folder is then a fully functional `electron-forge` project, so if you `cd electron-app` and then `yarn start` or yarn `electron-forge start`, it will launch your app just using `electron-forge` and no Ember tooling at all.

With this as a baseline, you can go find any tooling you like to auto-restart the Electron main process on file changes such as [nodemod](https://github.com/remy/nodemon) or [electron-reloader](https://github.com/sindresorhus/electron-reloader) and set it up in your `electron-app` project. In this mode, though, you don't get Ember's auto-reload functionality, because that requires `ember-cli`'s express server, etc. etc., but if you use `electron-reloader`, then if you leave an `ember electron:build --watch` process running in the background, it will rebuild Ember source changes into `electron-app/dist`, which will be noticed by `electron-reloader` to reload the browser window.

The TL;DR here is that this really isn't `ember-electron`'s job because it delegates all main process handling to `electron-forge`. But of course we're always open to ideas to make the integration easier for scenarios we haven't focused on.

**Comment by anulman** [comment]
@bendemboski That sounds like an excellent solution and recommendation 😄
